pkg/controller/security/manager.go

```diff
@@ -0,0 +1,289 @@
+			return heap.Remove(pq, i).(*certExp)
+		}
+	}
+	return nil
+}
+
+// Find the top item and pop if it is about to expire
+func (pq *certificateQueue) lookupAndPop() *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	top := pq.certs[0]
+	if time.Until(top.exp.Add(-10*time.Minute)) <= 0 {
+		return heap.Pop(pq).(*certExp)
+	}
+	return nil
+}
+
+func (s *SecretManager) lookupAndStoreCert(identity string, newCert *istiosecurity.SecretItem) {
+	s.certsCache.mu.Lock()
+	defer s.certsCache.mu.Unlock()
+	// Check if the key exists in the map
+	// If refCnt == 0, then this certificate is about to be deleted, so do not perform a refresh.
+	certCache := s.certsCache.certs[identity]
+	if certCache != nil && certCache.refCnt != 0 {
+		certCache.cert = newCert
+		certExp := certExp{exp: newCert.ExpireTime, identity: identity}
+		s.certsQueue.addItem(&certExp)
+		log.Infof("cert %v add, exp:%v\n", identity, newCert.ExpireTime)
+	}
+}
+
+// Check if the certificate already exists, if it does, increment the reference count by 1,
+// if it doesn't, request a new certificate.
+func (c *certsCache) lookupOrStore(identity string) *certCache {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	cert := c.certs[identity]
+	if cert != nil {
+		cert.refCnt++
+		return cert
+	}
+	cert = &certCache{
+		refCnt: 1,
+	}
+	c.certs[identity] = cert
+	return nil
+}
+
+func (c *certsCache) delete(identity string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	delete(c.certs, identity)
+}
+
+// NewSecretManager creates a new secretManager.s
+func NewSecretManager() (*SecretManager, error) {
+	tlsOpts := &tlsOptions{
+		RootCert: constants.RootCertPath,
+	}
+
+	options := NewSecurityOptions()
+	caClient, err := newCaClient(options, tlsOpts)
+	if err != nil {
+		return nil, err
+	}
+	certCache := newCertCache()
+	pq := newPriorityQueue()
+	heap.Init(pq)
+	certOpChan := make(chan securityData, maxGoroutines)
+
+	secretManager := SecretManager{
+		caClient:      caClient,
+		configOptions: options,
+		certsCache:    certCache,
+		certsQueue:    pq,
+		certsOpChan:   certOpChan,
+	}
+	go secretManager.operateCerts()
+	go secretManager.refreshExpiringCerts()
+	return &secretManager, nil
+}
+
+// Automatically check and refresh when the validity period expires
+// Store the identity in the priority queue according to the expiration time.
+// Check the highest priority element in the queue every 5 minutes.
+// If it is about to expire, pop up the element and reapply for the certificate.
+func (s *SecretManager) refreshExpiringCerts() {
+	for {
+		if s.certsQueue.Len() != 0 {
+			// As long as the memory in the Go language is referenced by any pointer,
+			// it will not be released. The map stores pointers, so it is safe here.
+			top := s.certsQueue.lookupAndPop()
+			if top != nil {
+				log.Debugf("refresh identity: %v  exp: %v\n", top.identity, top.exp)
+				newCert, err := s.caClient.fetchCert(top.identity)
+				if err != nil {
+					log.Errorf("%v refresh fetchCert error : %v", top.identity, err)
```

Comment:
shouldn't re-eneueue it?

Reply:
The identity of the workload may cannot be apply for a certificate. If the identity is re-entered to the queue, an infinite loop may occur.
It is best to only retry a finite number of times a finite number of times